litex/soc/software/include/hw/common.h

```diff
@@ -14,37 +14,234 @@
+static inline void csr_wr_uint8(uint8_t v, unsigned long a)
 {
-	*((volatile uint16_t *)addr) = value;
+	_csr_wr((unsigned long *)a, v, sizeof(uint8_t));
 }
 
-static inline uint16_t csr_readw(unsigned long addr)
+static inline uint16_t csr_rd_uint16(unsigned long a)
 {
-	return *(volatile uint16_t *)addr;
+	return _csr_rd((unsigned long *)a, sizeof(uint16_t));
 }
 
-static inline void csr_writel(uint32_t value, unsigned long addr)
+static inline void csr_wr_uint16(uint16_t v, unsigned long a)
 {
-	*((volatile uint32_t *)addr) = value;
+	_csr_wr((unsigned long *)a, v, sizeof(uint16_t));
 }
 
-static inline uint32_t csr_readl(unsigned long addr)
+static inline uint32_t csr_rd_uint32(unsigned long a)
 {
-	return *(volatile uint32_t *)addr;
+	return _csr_rd((unsigned long *)a, sizeof(uint32_t));
 }
+
+static inline void csr_wr_uint32(uint32_t v, unsigned long a)
+{
+	_csr_wr((unsigned long *)a, v, sizeof(uint32_t));
+}
+
+static inline uint64_t csr_rd_uint64(unsigned long a)
+{
+	return _csr_rd((unsigned long *)a, sizeof(uint64_t));
+}
+
+static inline void csr_wr_uint64(uint64_t v, unsigned long a)
+{
+	_csr_wr((unsigned long *)a, v, sizeof(uint64_t));
+}
+
+/* Read a CSR located at address 'a' into an array 'buf' of 'cnt' elements
+ * of type 'uintX_t'.
+ *
+ * NOTE: Since CSR_DW_BYTES is a constant here, we might be tempted to further
+ * optimize things by leaving out one or the other of the if() branches below,
+ * depending on each unsigned type width;
+ * However, this code is also meant to serve as a reference for how CSRs are
+ * to be manipulated by other programs (e.g., an OS kernel), which may benefit
+ * from dynamically handling multiple possible CSR subregister data widths
+ * (e.g., by passing a value in through the Device Tree).
+ * Ultimately, if CSR_DW_BYTES is indeed a constant, the compiler should be
+ * able to determine on its own whether it can automatically optimize away one
+ * of the if() branches! */
+#define _csr_rd_buf(a, uintX_t, buf, cnt) \
+{ \
+	int i, j, nsubs, n_sub_elem; \
+	unsigned long *addr = (unsigned long *)(a); \
+	uint64_t r; \
+	if (sizeof(uintX_t) >= CSR_DW_BYTES) { \
+		/* one or more subregisters per element */ \
+		for (i = 0; i < cnt; i++) { \
+			buf[i] = _csr_rd(addr, sizeof(uintX_t)); \
+			addr += num_subregs(sizeof(uintX_t)); \
+		} \
+	} else { \
+		/* multiple elements per subregister (2, 4, or 8) */ \
+		nsubs = num_subregs(sizeof(uintX_t) * cnt); \
+		n_sub_elem = CSR_DW_BYTES / sizeof(uintX_t); \
+		for (i = 0; i < nsubs; i++) { \
+			r = addr[i]; \
+			for (j = n_sub_elem - 1; j >= 0; j--) { \
+				if (i * n_sub_elem + j < cnt) \
+					buf[i * n_sub_elem + j] = r; \
+				r >>= sizeof(uintX_t) * 8; \
+			} \
+		} \
+	} \
+}
+
+/* Write an array 'buf' of 'cnt' elements of type 'uintX_t' to a CSR
+ * located at address 'a'.
+ *
+ * NOTE: The same optimization considerations apply here as with _csr_rd_buf()
+ * above.
+ */
+#define _csr_wr_buf(a, uintX_t, buf, cnt) \
```

Comment:
Is there any reason this whole thing should be a macro? It looks like you just use `sizeof(type uintX_t`)? 

I would expect that if this was an `inline` function then it would end up as the same ASM?

Reply:
There's also `buf[i]`, which is of type `uintX_t`. I'd have to have some sort of switch statement and a bunch of casts to each actual width (`X`), and make `buf` a `void *`. I figured the loss of legibility wasn't worth it, but if there's a nice elegant way to still do an inline function instead of a macro that'd be awesome!

Reply:
... but at least I might be able to avoid passing `uintX_t` to the macro, and use `sizeof(buf[0])` instead... I'll take another look after I get some sleep, and force-push a cleaned-up version if it still makes sense in the morning :)

Reply:
force-pushed a simplified pair of macros, without the `uintX_t` parameter, which is implicit in the type of `buf[0]` (tested, and working same as before). I think we're stuck with the macros, since implementing the casts implied by the type of `buf[0]` manually would be much less readable, IMHO.